src/context/backward_tokens.rs

```diff
@@ -52,22 +53,27 @@ macro_rules! move_to_prev_sibling_or_try_go_to_parent_and_loop {
     };
 }
 
-pub fn get_backward_tokens(node: Node) -> impl Iterator<Item = Node> {
-    BackwardTokenWalker::new(node)
+pub fn get_backward_tokens<'a>(
+    node: Node<'a>,
+    node_parent_provider: StandaloneNodeParentProvider<'a>,
+) -> impl Iterator<Item = Node<'a>> {
+    BackwardTokenWalker::new(node, node_parent_provider)
 }
 
 struct BackwardTokenWalker<'a> {
     state: TokenWalkerState,
     node: Node<'a>,
     original_node: Node<'a>,
+    node_parent_provider: StandaloneNodeParentProvider<'a>,
```

Comment:
There were places where having another lifetime (eg `'b` on `QueryMatchContext<'a, 'b>`) would have been not helpful so sort of "exposed" a `StandaloneNodeParentProvider` struct that can be acquired from any implementor of `NodeParentProvider` and doesn't have any additional lifetime dependencies